### PR TITLE
Practice pull request

### PR DESCRIPTION
**Pull request by stephan-koenig**
This is the basic introduction to R and RStudio tutorial.

### CODE REVIEW

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+### Loading packages
+
+After installing a package, and *everytime* you open a new RStudio session, you need to first load (open) the packages you want to use with the `library()` function. This tells R to access the package's functions and prevents RStudio from lags that would occur if it automatically loaded every downloaded package every time you opened it.
+
+Packages can be loaded like this:
+
+```{r eval = FALSE}
+library(tidyverse)
+```
+
```
Comment (r-karimi) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:171]: It is a little tricker to load Bioconductor packages, as they are often not stored on the Comprehensive R Archive Network (CRAN) where most packages live. There is a package, however, that lives on CRAN and serves as an interface between CRAN and the Bioconductor packages.

To load a Bioconductor package, you must first install and load the BiocManager package, like so.

`install.packages("BiocManager")`
`library("BiocManager`)

You can then use the function `BiocManager::install()` to install a package. To install the `Annotate` package, we would execute the following code.

`BiocManager::install("annotate")`

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 2]
```diff
@@ -0,0 +1,284 @@
+---
+title: "Introduction to R and RStudio fundamentals"
+author: "Michelle Kang (adapted from Dr. Kim Dill-McFarland)"
+date: "version `r format(Sys.time(), '%B %d, %Y')`"
+output:
+  learnr::tutorial:
+    progressive: true
+    allow_skip: true
+runtime: shiny_prerendered
+description: In-depth description will follow.
```
Comment (r-karimi) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:10]: This tutorial covers the basics of R and RStudio. You will learn about the different panes and features of RStudio that make coding in R easier, as well as basic skills of the R language itself, such as creating functions and loading packages.

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+The **Usage** section tells us the inputs that need to be specified and default inputs of read_delim:
+
+- `file` and `delim` need to be specified as they are not followed by `=`
+- all other parameters have a default value e.g. `quote = "\"` and do not have to be specified to run the function.
+
+The **Arguments** Section describes the requirements of each input argument in detail.   
+
+The **Examples** Section has examples of the function that can be directly copy and pasted into your terminal and ran. 
+
+
```
Comment (psampara) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:272]: Another example from base R that may be widely used is the function `nrow()` 

```{r eval = FALSE}
?nrow
```

The **Description** section tells us that `nrow()` from a matrix or an array.

The **Usage** section tells us the inputs that need to be specified and default inputs of `nrow()`:

- `x` is the data matrix or array for which the user is interested in identifying the number of rows

The **Arguments** Section describes the requirements of each input argument in detail.

The **Examples** Section has examples of the function that can be directly copy and pasted into your terminal and ran.



Tidyverse is a wrapper for many valuable functions widely used in R. One of the examples from Tidyverse would be `select()`

```{r eval = FALSE}
if (!require("tidyverse")) install.packages("tidyverse")
library(tidyverse)
?select
```

The **Description** section tells us that `select()` can be used to select certain columns from a parent dataset, or optionally rename the columns

The **Overview of selection features** section provides the user a list of operators and selection helpers to fully realize the power of `select()` 

The **Usage** section tells us the inputs that need to be specified and default inputs of `select()`

- `.data` is a mandatory input which refers to the parent dataset or tibble to subset from
- the helper functions and list of operators could be used to specify the columns to subset

The **Value** section describes an object integral to `select()`. A more descriptive account can be read in the help section

The **Method** section describes the implementation method of the function `select()`

The **Examples** Section has examples of the function that can be directly copy and pasted into your terminal and ran.

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+By the end of this tutorial you should be able to:
+
+- Identify the different components of RStudio.
+- Declare variables in R.
+- Recognize and use functions.
+- Install and load R packages.
+- Load and subset tabular data using tidyverse.
+- Use the `help` function in R console to troubleshoot given a new function.
+
```
Comment (psampara) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:36]:
```suggestion
The last bullet point can be more descriptive saying, "Use the `help` function in R console to troubleshoot and *identify mandatory parameters* given a new function."
```
Other than that, the list looks comprehensive.

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+- `file` and `delim` need to be specified as they are not followed by `=`
+- all other parameters have a default value e.g. `quote = "\"` and do not have to be specified to run the function.
+
+The **Arguments** Section describes the requirements of each input argument in detail.   
+
+The **Examples** Section has examples of the function that can be directly copy and pasted into your terminal and ran. 
+
+
+
```
Comment (psampara) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:273]:
```suggestion
Could add a couple of questions to make the tutorial more interactive

```{r quiz: navigating help, echo=FALSE}
question("How would you launch a help section for a given function",
  answer("Place an exclamation mark in front of the function"),
  answer("Place an question mark in front of the function", correct = TRUE),
  answer("Place an question mark after of the function"),
  answer("Type out help and function name in the console")
)

question("What does an = sign indicate in the help section",
         answer("There exists a default and is not mandatory", correct = TRUE),
         answer("Is a mandatory input in the function")
)
```
```

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+## Variables in R
+
+We use variables to store data that we want to access or manipulate later. Variables must have unique names.  
+
+Without declaring a variable the sum of these two numbers will be printed to console but cannot be accessed for future use:
+
+```{r}
```
Comment (ymkng) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:87]:
```suggestion
```{r novar, exercise=TRUE}
```

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+2 + 2 
+```
+
+To declare a variable, follow the pattern of: `variable <- value`. Let's declare a variable `total` as the sum of two numbers.
+
+```{r}
```
Comment (ymkng) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:93]:
```suggestion
```{r d_var, exercise=TRUE}
```

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+total <- 2 + 2
+```
+
+We access the value of `total`:
+
+```{r}
```
Comment (ymkng) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:99]:
```suggestion
```{r var, exercise=TRUE}
```

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+total
+```
+
+We can use the value stored in `total`:
+
+```{r}
```
Comment (ymkng) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:105]:
```suggestion
```{r sub_var, exercise=TRUE}
```

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+total - 1
+```
+
+After declaring a variable, we can perform operations to change the value stored in the variable:
+
+```{r}
```
Comment (ymkng) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:111]:
```suggestion
```{r sub_var2, exercise=TRUE}
```

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+total <- total - 1
+```
```
Comment (ymkng) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:113]:
```suggestion

```

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+
+Now `total` stores a new value:
```
Comment (ymkng) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:115]:
```suggestion

```

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+
+```{r}
```
Comment (ymkng) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:117]:
```suggestion

```

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+# General learnr setup
+library(learnr)
+knitr::opts_chunk$set(echo = TRUE)
+library(educer)
+# Helper function to set path to images to "/images" etc.
+setup_resources()
+
+# Tutorial specific setup
+library(dplyr)
+library(readr)
```
Comment (ymkng) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:23]:
```suggestion
library(readr)
total <- 4
```

**Review by ymkng — COMMENTED** [review]
Makes variable section interactive

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd (hunk 2 above)
Comment (r-karimi) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:10]:
```suggestion
description: This tutorial covers the basics of R and RStudio. You will learn about the different panes and features of RStudio that make coding in R easier, as well as basic skills of the R language itself, such as creating functions and loading packages.
```

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd (hunk 1 above)
Comment (r-karimi) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:171]:
```suggestion
It is a little tricker to load Bioconductor packages, as they are often not stored on the Comprehensive R Archive Network (CRAN) where most packages live. There is a package, however, that lives on CRAN and serves as an interface between CRAN and the Bioconductor packages.

To load a Bioconductor package, you must first install and load the BiocManager package, like so.

`install.packages("BiocManager")`
`library("BiocManager)`

You can then use the function `BiocManager::install()` to install a Bioconductor package. To install the Annotate package, we would execute the following code.

`BiocManager::install("annotate")`
```

**Review by r-karimi — COMMENTED** [review]
Revise incorrect previous code review

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+
+```{r}
+total
+```
+
+
+
+## Functions
+
+Functions are one of the basic units in programming. Generally speaking, a function takes some input and generates some output. Every R function follows the same basic syntax, where `function()` is the name of the function and `arguments` are the different parameters you can specify (i.e. your input):
```
Comment (Noonanav) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:125]:
```suggestion
Functions are one of the basic units in programming. Generally speaking, a function takes some input and generates some output, in a reproducible way. Every R function follows the same basic syntax, where `function()` is the name of the function and `arguments` are the different parameters you can specify (i.e. your input):
```

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+
+`function(argument1 = ..., argument2 = ..., ...)`
+
+You can treat functions as a black box and do not necessarily need to know how it works under the hood as long as your provided input conforms to a specific shape.
```
Comment (Noonanav) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:129]:
```suggestion
You can treat functions as a black box and do not necessarily need to know how it works under the hood as long as your provided input conforms to a specific format.
```

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+
+![](/images/function.png){width=75%}
+
+For example, the function `sum()` expects numbers:
```
Comment (Noonanav) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:133]:
```suggestion
For example, the function `sum()` (which outputs the sum of the arguments) expects numbers:
```

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+
+```{r}
```
Comment (Noonanav) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:135]:
```suggestion
```{r sum_function, exercise=TRUE}
```

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+sum(3, 5, 9, 18)
+```
+
+If you instead pass text as arguments to `sum()` you will receive an error:
+
+```{r error = TRUE}
```
Comment (Noonanav) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:141]:
```suggestion
```{r sum_text, exercise=TRUE, error = TRUE}
```

File: inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+sum("Sum", "does", "not", "accept", "text!")
+```
+
```
Comment (Noonanav) [inline inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:144]:
```suggestion
The use of functions isn't limited to mathematical calculations. Function can also be used to transform data.

For example, the function `t()` can be used to transpose a matrix. In the example below, we generate a matrix (`example_matrix`) using the functions `c()` and `as.matrix()`, which were used to combine values into a vector and transform the vector into a matrix. We then use the `t()` function to transpose `example_matrix` into the matrix `example_transposed`.

```{r t_function, exercise=TRUE}
# Generate Matrix
example_list <- c("T", "does", "accept", "text!")
example_matrix <- as.matrix(example_list)

#Transpose Matrix
example_transposed <- t(example_matrix)

#Display Original and Transposed Matrices
example_matrix
example_transposed
```
### The most helpful function of all: the 'help' function
You can get information about a specific function by running the command `?<function>` or `help(<function>)` (replace `<function>` by the name of the function you are interested in). This command opens the help page, where you can find all information about a function's purpose and its arguments. For beginners, it is useful to concentrate on the "Examples" and "Arguments" section to understand the typical usage of the function better.

Run the code below to read the documentation for the `t()` function.

```{r transpose_help, exercise=TRUE}
?t
```
```

Comment (Noonanav) [reply inst/tutorials/r_and_rstudio_basic/r_and_rstudio_basic.Rmd:144]: I copied this code over from my .Rmd file, but it seems like the "suggested change" syntax interferes with the RMarkdown syntax.  Will this be a problem when we accept the changes?

**Review by Noonanav — COMMENTED** [review]
I have updated the Functions section of the document according to our conversations last week

**Pull request closed without merging**